From 4c035eaf1d59d1df9a09285b36c9b9b1b90764e4 Mon Sep 17 00:00:00 2001
From: Simon Busch <morphis@gravedo.de>
Date: Mon, 5 Aug 2013 10:39:21 +0000
Subject: [PATCH 2/2] Add support for systemd

In order to synchronize our service launch on boot we need to add explicit support for
systemd to notify systemd when we're really set up and other depending services can be
launched. As ls-hubd is a critical component of the system it has to be really up before
anyone can use it in order not to fail.

Upstream-Status: pending (http://github.com/openwebos/luna-service2)

Open-webOS-DCO-1.0-Signed-off-by: Simon Busch <morphis@gravedo.de>
---
 CMakeLists.txt             | 4 ++++
 src/ls-hubd/CMakeLists.txt | 2 +-
 src/ls-hubd/hub.c          | 6 ++++++
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index cc4e243..081e601 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,6 +45,10 @@ pkg_check_modules(PMLOGLIB REQUIRED PmLogLib)
 include_directories(${PMLOGLIB_INCLUDE_DIRS})
 webos_add_compiler_flags(ALL ${PMLOGLIB_CFLAGS_OTHER})
 
+pkg_check_modules(SYSTEMD_DAEMON REQUIRED libsystemd-daemon)
+include_directories(${SYSTEMD_DAEMON_INCLUDE_DIRS})
+webos_add_compiler_flags(ALL ${SYSTEMD_DAEMON_CFLAGS_OTHER})
+
 webos_machine_impl_dep()
 include(webOS/LegacyDefines)
 
diff --git a/src/ls-hubd/CMakeLists.txt b/src/ls-hubd/CMakeLists.txt
index 3dde602..738b8c0 100644
--- a/src/ls-hubd/CMakeLists.txt
+++ b/src/ls-hubd/CMakeLists.txt
@@ -35,6 +35,6 @@ add_definitions(-DDEFAULT_PID_DIR="${WEBOS_INSTALL_RUNTIMEINFODIR}/ls2")
 add_definitions(-DDEFAULT_LOCAL_SOCKET_DIR="${WEBOS_INSTALL_RUNTIMEINFODIR}/ls2")
 
 add_executable(ls-hubd ${HUB_SOURCE_FILES})
-target_link_libraries(ls-hubd ${CMAKE_PROJECT_NAME} ${PMLOGLIB_LDFLAGS})
+target_link_libraries(ls-hubd ${CMAKE_PROJECT_NAME} ${PMLOGLIB_LDFLAGS} ${SYSTEMD_DAEMON_LDFLAGS})
 webos_build_daemon(NAME ls-hubd LAUNCH ${CMAKE_SOURCE_DIR}/files/launch/ ${LS2_RESTRICTED})
 
diff --git a/src/ls-hubd/hub.c b/src/ls-hubd/hub.c
index 1d19bd5..e3903ab 100644
--- a/src/ls-hubd/hub.c
+++ b/src/ls-hubd/hub.c
@@ -29,6 +29,7 @@
 #include <sys/stat.h>
 #include <libgen.h>
 #include <glib.h>
+#include <systemd/sd-daemon.h>
 
 #include "hub.h"
 #include "conf.h"
@@ -4266,6 +4267,11 @@ main(int argc, char *argv[])
             g_critical("Unable to emit upstart event");
         }
     }
+
+    if (getenv("NOTIFY_SOCKET") != NULL)
+    {
+        sd_notify(0, "READY=1");
+    }
 #endif
     
     if (boot_file_name)
-- 
1.8.1.2

